src/main/java/org/primefaces/component/datatable/DataTable.java

```diff
@@ -1542,11 +1527,152 @@ public boolean isColumnSortable(UIColumn column) {
+            filterByTmp = Collections.singletonList((FilterMeta) usrFilterBy);
+        }
+        else if (!(usrFilterBy instanceof Collection)) {
+            throw new FacesException("DataTable#filterBy expects a single or a collection of FilterMeta");
+        }
+        else {
+            filterByTmp = (Collection<FilterMeta>) usrFilterBy;
+        }
+
+        for (FilterMeta userFM : filterByTmp) {
+            FilterMeta intlFM = intlFilterBy.values().stream()
+                    .filter(o -> o.getField().equals(userFM.getField()))
+                    .findAny()
+                    .orElseThrow(() -> new FacesException("No column with field '" + userFM.getField() + "' has been found"));
+
+            ValueExpression filterByVE = userFM.getFilterBy();
+            if (filterByVE == null) {
+                filterByVE = createValueExprFromVarField(getFacesContext(), getVar(), userFM.getField());
+            }
+
+            intlFM.setFilterValue(userFM.getFilterValue());
+            intlFM.setFilterBy(filterByVE);
+            intlFM.setConstraint(userFM.getConstraint());
+            intlFM.setMatchMode(userFM.getMatchMode());
+            filtered.set(filtered.get() || userFM.isActive());
+        }
+    }
+
+    protected void updateFilterByWithGlobalFilter(Map<String, FilterMeta> filterBy, AtomicBoolean filtered) {
+        String globalFilter = getGlobalFilter();
+        Set<SearchExpressionHint> hint = LangUtils.isValueBlank(globalFilter)
+                ? EnumSet.of(SearchExpressionHint.IGNORE_NO_RESULT)
+                : Collections.emptySet();
+        UIComponent globalFilterComponent = SearchExpressionFacade
+                .resolveComponent(getFacesContext(), this, DataTable.PropertyKeys.globalFilter.toString(), hint);
+        if (globalFilterComponent != null) {
+            if (globalFilterComponent instanceof ValueHolder) {
+                ((ValueHolder) globalFilterComponent).setValue(globalFilter);
+            }
+            FilterMeta globalFilterBy = FilterMeta.of(filterBy.values(), globalFilter, getGlobalFilterFunction());
+            filterBy.put(globalFilterBy.getColumnKey(), globalFilterBy);
+            filtered.set(filtered.get() || globalFilterBy.isActive());
+        }
+    }
+
+    public boolean isColumnFilterable(UIColumn column) {
+        Map<String, FilterMeta> filterBy = getFilterByAsMap();
+        return filterBy.containsKey(column.getColumnKey());
+    }
+
     public Map<String, SortMeta> getSortByAsMap() {
-        return ComponentUtils.computeIfAbsent(getStateHelper(), "_sortBy", () -> initSortBy(getSortBy()));
+        return ComponentUtils.computeIfAbsent(getStateHelper(), "_sortBy", this::initSortBy);
     }
 
     public void setSortByAsMap(Map<String, SortMeta> sortBy) {
         getStateHelper().put("_sortBy", sortBy);
     }
+
+    public Map<String, FilterMeta> getFilterByAsMap() {
+        return ComponentUtils.eval(getStateHelper(), "_filterBy", Collections::emptyMap);
```

Comment:
This will add an unmodifiable map so the user won't be able to add values to it.  Should `new HashMap<>` be used here?

Reply:
I could have returned null, but I don't like returning null for collections (so I don't have to test for nullity afterwards), also users are not supposed to access/modify this map as it is used internally